File: src/tutor/java-guide-source-reader.js
import { readFile } from "node:fs/promises";
import path from "node:path";
import { safeSnippet, toReadableSourceText } from "../material/material-model.js";

const defaultJavaGuideRoot = process.env.LLAI_JAVAGUIDE_ROOT || "/Users/lee/IdeaProjects/JavaGuide";

function resolveSourcePath(sourcePath) {
  return path.join(defaultJavaGuideRoot, sourcePath);
}

export async function loadJavaGuideSourceSnippets(sources = [], limit = 2) {
  const selected = sources.slice(0, limit);
  const loaded = await Promise.all(
    selected.map(async (source) => {
      try {
        const raw = await readFile(resolveSourcePath(source.path), "utf8");
        const normalized = toReadableSourceText(raw);
        return {
          path: source.path,
          title: source.title,
          url: source.url,
          snippet: safeSnippet(normalized, 900)
        };
      } catch {
        return {
          path: source.path,
          title: source.title,
          url: source.url,
          snippet: ""
        };
      }
    })
  );

  return loaded.filter((item) => item.title);
}

File: ai-service/app/engine/java_guide_source_reader.py
from __future__ import annotations

import os
from pathlib import Path
from typing import Any, Dict, List


DEFAULT_JAVA_GUIDE_ROOT = os.environ.get("LLAI_JAVAGUIDE_ROOT", "/Users/lee/IdeaProjects/JavaGuide")


def _safe_snippet(text: str, limit: int = 900) -> str:
    normalized = " ".join(str(text or "").split()).strip()
    if len(normalized) <= limit:
        return normalized
    return f"{normalized[: max(0, limit - 1)].strip()}…"


def load_java_guide_source_snippets(sources: List[Dict[str, Any]] | None = None, limit: int = 2) -> List[Dict[str, Any]]:
    selected = (sources or [])[:limit]
    loaded: List[Dict[str, Any]] = []
    for source in selected:
        source_path = Path(DEFAULT_JAVA_GUIDE_ROOT) / str(source.get("path", ""))
        snippet = ""
        try:
            raw = source_path.read_text(encoding="utf-8")
            snippet = _safe_snippet(raw, 900)
        except Exception:
            snippet = ""
        loaded.append(
            {
                "path": source.get("path", ""),
                "title": source.get("title", ""),
                "url": source.get("url", ""),
                "snippet": snippet,
            }
        )
    return [item for item in loaded if item.get("title")]
